overlord/hookstate/ctlcmd/refresh.go

```diff
@@ -122,33 +125,68 @@ type updateDetails struct {
 	restart bool
 }
 
-func getUpdateDetails(context *hookstate.Context) *updateDetails {
+// refreshCandidate is a subset of refreshCandidate defined by snapstate and
+// stored in "refresh-candidates".
+type refreshCandidate struct {
+	Channel     string         `json:"channel,omitempty"`
+	Version     string         `json:"version,omitempty"`
+	SideInfo    *snap.SideInfo `json:"side-info,omitempty"`
+	InstanceKey string         `json:"instance-key,omitempty"`
+}
+
+func getUpdateDetails(context *hookstate.Context) (*updateDetails, error) {
 	context.Lock()
 	defer context.Unlock()
 
 	if context.IsEphemeral() {
 		// TODO: support ephemeral context
-		return nil
+		return nil, nil
 	}
 
 	var base, restart bool
 	context.Get("base", &base)
 	context.Get("restart", &restart)
 
-	// TODO: get revision, version etc. from refresh-candidates.
+	var candidates map[string]*refreshCandidate
+	st := context.State()
+	if err := st.Get("refresh-candidates", &candidates); err != nil {
+		return nil, err
+	}
 
+	// TODO: pending
 	up := updateDetails{
 		base:    base,
 		restart: restart,
 	}
-	return &up
+
+	// try to find revision/version/channel info from refresh-candidates; it
+	// may be missing if the hook is called for snap that is just affected by
+	// refresh but not refreshed itself, in such case this data is not
+	// displayed.
+	if cand, ok := candidates[context.InstanceName()]; ok {
+		up.channel = cand.Channel
+		up.revision = cand.SideInfo.Revision
+		up.version = cand.Version
+		return &up, nil
+	}
+
+	// refresh-hint not present, look up channel info in snapstate
+	var snapst snapstate.SnapState
+	if err := snapstate.Get(st, context.InstanceName(), &snapst); err != nil {
+		return nil, fmt.Errorf("internal error: cannot get snap state for %q: %v", context.InstanceName(), err)
+	}
+	up.channel = snapst.TrackingChannel
+	return &up, nil
 }
 
-func (c *refreshCommand) printPendingInfo() {
-	details := getUpdateDetails(c.context())
+func (c *refreshCommand) printPendingInfo() error {
+	details, err := getUpdateDetails(c.context())
+	if err != nil {
+		return err
+	}
 	// XXX: remove when ephemeral context is supported.
 	if details == nil {
-		return
+		return nil
 	}
 	c.printf("pending: %s\n", details.pending)
 	c.printf("channel: %s\n", details.channel)
```

Comment:
I forgot about this but version might confuse things  (1.0 vs "1.0"), I wonder if we should use yaml.Marshal here instead of multiple printf.

Reply:
That's probably a good idea (OTOH we don't seem to be using it anywhere for similar cases, e.g. snap info).
I've done this change, but wondering a bit about revision, currently forced a number there instead of snap.Revision because it would print e.g. "3". (but also "unset" for rev=0).